### PR TITLE
Adding the Fully Parametrized Quantile Network

### PR DESCRIPTION
**Pull request by apprenticearnab**

### CODE REVIEW

**Comment by findmyway** [comment]
Hi @apprenticearnab ,

Could you fix the test error first? I'll look into the implementation details in the next week.

**Comment by apprenticearnab** [comment]
Hi , I am facing problem while integrating the developed package with the ReinforcementLearning.jl package , it will be helpful if anybody please suggest any alternative way for running the experiments without integrating with ReinforcementLearning.jl package , or the correct way to integrate the developed package as you can see in the commits with the ReinforcementLearning.jl package . I am using this type of configuration for the learner i.e. FPNLearner in this case : 
`mutable struct FPNLearner{A <: AbstractApproximator,T <: AbstractApproximator,B <: NeuralNetworkApproximator,R,D} <: AbstractLearner `. Also , please suggest if this type of configuration for AbstractLearner is feasible or not .

**Comment by findmyway** [comment]
For local development, you can try:

```julia
using ReinforcementLearningBase, ReinforcementLearningCore, ReinforcementLearningEnvironments, ReinforcementLearningZoo
```

Note that `ReinforcementLearningEnvironments` is not a direct dependency. I usually have a test folder to `]dev /path/to/ReinforcementLearningZoo.jl` and `]add ReinforcementLearningEnvironments`. Then make tests there.

**Comment by apprenticearnab** [comment]
Please help me in fixing this issue ,
```
no method matching update!(::FPNLearner{NeuralNetworkApproximator{FullyParametrizedNet{Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}},ADAM},NeuralNetworkApproximator{FullyParametrizedNet{Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},Dense{typeof(identity),Array{Float32,2},Array{Float32,1}}},Nothing},NeuralNetworkApproximator{Dense{typeof(relu),Array{Float32,2},Array{Float32,1}},RMSProp},StableRNGs.LehmerRNG}, ::PrioritizedTrajectory{Trajectory{NamedTuple{(:state, :action, :reward, :terminal),Tuple{CircularArrayBuffers.CircularArrayBuffer{Float32,2},CircularArrayBuffers.CircularArrayBuffer{Int64,1},CircularArrayBuffers.CircularArrayBuffer{Float32,1},CircularArrayBuffers.CircularArrayBuffer{Bool,1}}}},SumTree{Float32}})
  Closest candidates are:
    update!(!Matched::RandomStartPolicy, ::Any) at C:\Users\alipu\.julia\packages\ReinforcementLearningBase\f5BSO\src\implementations\policies.jl:99
    update!(!Matched::RandomPolicy, ::Any) at C:\Users\alipu\.julia\packages\ReinforcementLearningCore\XZfCY\src\extensions\ReinforcementLearningBase.jl:7
```
I tried different approaches but failed !!

**Comment by findmyway** [comment]
You might need to spend some time on how each `update!` function is called:

```
agent(env) ->
update!(agent, trajectory) -> 
update!(policy, trajectory) -> 
update!(learner, trajectory) -> 
update!(learner, NamedTuple) ->
update!(approximator, gradients)
```

So the error you pasted above is caused by that you haven't implemented `update!(learner, trajectory)` here.

File: src/algorithms/dqns/fpqn.jl [hunk 1]
```diff
@@ -0,0 +1,330 @@
...
+           ↑
+         header
+           ↑
+feature ↱  ⨀   ↰ transformed embedding
+       ψ       ϕ
+       ↑       ↑
+       s        τ
+```
+"""
+Base.@kwdef struct FullyParametrizedNet{A,B,C} <: AbstractApproximator
```
Comment (findmyway) [inline src/algorithms/dqns/fpqn.jl:24]: So what's the difference between `FullyParametrizedNet` and `ImplicitQuantileNet`?

Comment (apprenticearnab) [reply src/algorithms/dqns/fpqn.jl:24]: The main difference lies in the learner only , as the fraction proposal network is a simple `NeuralNetworkApproximator`

Comment (findmyway) [reply src/algorithms/dqns/fpqn.jl:24]: If so, can we just reuse `ImplicitQuantileNet`?

Comment (apprenticearnab) [reply src/algorithms/dqns/fpqn.jl:24]: Yes , we can reuse that

**Pull request closed without merging**